File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

const int MX = 20002;
bool visited[MX];
int dist[MX];

vector<int> vec[MX];

void bfs(){
    queue<int> q;
    q.push(1);
    visited[1] = true;
    
    int d = 0;
    while(!q.empty()){
        int sz = q.size();
        while(sz--){
            int cur = q.front();
            q.pop();
            dist[cur] = d;
            for(int i = 0; i < vec[cur].size(); i++){
                int nxt = vec[cur][i];
                if(visited[nxt]) continue;
                q.push(nxt);
                visited[nxt]= true;
            }
        }
        d++;
    }
}

int solution(int n, vector<vector<int>> edge) {

    
    for(int i = 0; i < edge.size(); i++){
        int a = edge[i][0];
        int b = edge[i][1];
        vec[a].push_back(b);
        vec[b].push_back(a);
    }
    
    bfs();
    
    int mx = -1;
    for(int i = 0; i < MX; i++){
        mx = max(mx, dist[i]);
    }
    
    int cnt = 0;
    for(int i = 0; i < MX; i++){
        if(mx == dist[i]) cnt++;
    }
    
    return cnt;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    final int MX = 20012;
    List<Integer>[] vec = new ArrayList[MX];
    boolean ch[] = new boolean[MX];
    
    int bfs(){
        Queue<Integer> q = new ArrayDeque<>();
        
        q.add(1);
        ch[1] = true;
        int mx = -1;
        int cnt = 0;
        int d = 0;
        
        while(!q.isEmpty()){
            int sz = q.size();
         
            while(sz-- > 0){
                int cur = q.poll();
                // System.out.println(cur + " " + d);
                if(d > mx) {
                    cnt = 1;
                    mx = d;
                } else if(d == mx){
                    cnt++;
                }
                for(int nxt : vec[cur]) {
                    if(ch[nxt]) continue;
                    q.add(nxt);
                    ch[nxt] = true;
                }
            }
            d++;
        }
        
        return cnt;
    }

    public int solution(int n, int[][] edge) {
        for(int i = 0; i < MX; i++){
            vec[i] = new ArrayList<>();
        }
        Arrays.fill(ch, false); // 초기화

        for(int i = 0; i < edge.length; i++){ // 수정된 부분
            int a = edge[i][0];
            int b = edge[i][1];
            vec[a].add(b);
            vec[b].add(a);
        }
       
        return bfs();
    }
}
